hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java

```diff
@@ -4148,13 +4158,53 @@ private String substituteBrackets(String sql) {
 		return new SubstituteBracketSQLQueryParser( sql, getFactory() ).process();
 	}
 
+	private void resolveIdentityInsertDelayable() {
+		// By default they can
+		// The remainder of this method checks use cases where we shouldn't permit it.
+		canIdentityInsertBeDelayed = true;
+
+		if ( getEntityMetamodel().getIdentifierProperty().isIdentifierAssignedByInsert() ) {
+			// if the persister's identifier is assigned by insert, we need to see if we must force non-delay mode.
+			for ( NonIdentifierAttribute attribute : getEntityMetamodel().getProperties() ) {
+				final Type attributeType = attribute.getType();
+				if ( attributeType.isAssociationType() ) {
+					if ( attributeType.isEntityType() ) {
+						// In this use case, the association check is to see if we self-reference.
+						Class<?> entityClass = attributeType.getReturnedClass();
+						if ( getMappedClass().equals( entityClass ) ) {
+							canIdentityInsertBeDelayed = false;
+						}
```

Comment:
Is this a problem only when the attribute is actually a reference to the entity being persisted? 

What about an association within an embeddable that refers to the entity being persisted? It doesn't seem like the check would work for that use case.

What if the association is a reference to a different entity of the same class? Could/should it be possible to delay the insert in this case? I suppose that would require the persister to track which attributes (or sub-attributes) are of the same class, then they would have to be checked to see if any refer to the same entity being persisted. Maybe this could be an improvement fixed later?

Reply:
> Is this a problem only when the attribute is actually a reference to the entity being persisted?

Yes

> What about an association within an embeddable that refers to the entity being persisted? It doesn't seem like the check would work for that use case.

Good catch and this does not work; I'll fix that use case too.

> Maybe this could be an improvement fixed later?

I think so, that's the main reason why we decided to introduce the configuration parameter for situations where we may have missed a use case that users could report but have a way to work around the problem in the short-term until we could revise the criteria this method uses.